### DISCUSSION TITLE
yolov5/tutorials/model_export/

### DISCUSSION THREAD

**Discussion by giscus** [ultralytics/ultralytics#9558] [Docs]
# yolov5/tutorials/model_export/

Learn how to export a trained YOLOv5 model from PyTorch to different formats including TorchScript, ONNX, OpenVINO, TensorRT, and CoreML, and how to use these models.

https://docs.ultralytics.com/yolov5/tutorials/model_export/

**Comment by Halafz7** [comment]
Can I validate the model with different datasets? Let's say I trained the model using COCO and validated it using ImageNet?

**Reply by pderrenger** [reply]
Hey there! 😊

Yes, you can definitely validate your model with different datasets, like training on COCO and then validating on ImageNet. It's a great way to check how well your model generalizes to different data distributions. Just make sure to adjust your dataset paths and format accordingly in your validation dataset's YAML file.

For example, if you've trained your model on COCO and want to validate on ImageNet, you might do something like this:

```bash
yolo val model=yolov8n.pt data=imagenet.yaml
```

Remember to prepare the `imagenet.yaml` file to point to your ImageNet validation data.

Happy validating! 🚀

**Comment by iraa777** [comment]
Hello,
I am trying to use yolov5 on a NVIDIA jetson nano. I want to export a PyTorch model to TensorRT, and the code you provide will not work. Since I am working on a jetson nano, I can not install all the requirements you have specified at the given version (it is usually an older version). Do you have any export.py alternative code?

**Reply by iraa777** [reply]
I forgot to mention that i have jetpack 4.6.1 and I am using docker to o this with base image nvcr.io/nvidia/l4t-pytorch:r32.7.1-pth1.10-py3

**Reply by glenn-jocher** [reply]
👋 Hey, happy to help with your Jetson Nano TensorRT export issue! Given your Jetpack 4.6.1 and Docker setup with the base image `nvcr.io/nvidia/l4t-pytorch:r32.7.1-pth1.10-py3`, you're right; some of the direct `export.py` might not work seamlessly due to version constraints.

For a more Jetson-friendly approach to exporting YOLO models to TensorRT, you might consider a step-by-step process starting with ONNX as an intermediary format. Here's a streamlined version tailored for your setup:

1. **Export to ONNX:**
   First, convert the PyTorch model to ONNX. This step is generally more straightforward and less prone to compatibility issues.
   ```bash
   python3 export.py --weights yolov5s.pt --include onnx
   ```

2. **Use ONNX with TensorRT on Jetson:**
   Once you have the ONNX model, you can use NVIDIA's `trtexec` tool (available on Jetson) or write a custom script to convert the ONNX model to TensorRT. The `trtexec` tool might look something like this:
   ```bash
   trtexec --onnx=yolov5s.onnx --saveEngine=yolov5s.engine
   ```

3. **Docker Considerations:**
   Since you're using Docker, ensure that your container has access to the GPU. You might also need to install additional dependencies for ONNX and TensorRT conversion inside your Docker container if they're not already included in your Jetson's base image.

4. **TensorRT Version Compatibility:**
   Check the version of TensorRT supported by your JetPack version to avoid compatibility issues with the exported files. JetPack 4.6.1 typically comes with TensorRT 8.x. Ensure that your model export aligns with this TensorRT version for seamless execution.

Let me know if these steps help or if you encounter any specific errors along the way! 🚀

**Comment by LGCCoding** [comment]
Dear developers, hello, I have a module does not exist in the export of tflite format files, I modified the official network structure of yolov5s, add a new module, can not be exported to tflite format, I want to know how to solve this problem.

**Reply by pderrenger** [reply]
Hey there 👋! Thanks for reaching out.

For your custom module not exporting to TFLite format, it's generally due to operations in the new module that TFLite might not support directly.

A few things to try:
1. **Review TFLite's supported ops**: Ensure your custom module only uses operations supported by TFLite.
2. **Simplify the model**: Use `torch.nn` modules and functions known to convert well to TFLite.
3. **Test with `torch.jit.trace`**: Before exporting to TFLite, test the export with TorchScript using `torch.jit.trace`, ensuring it's successful.
4. **Use PyTorch's ` @LGCCoding.jit.ignore`**: For operations not crucial in inference, you can skip them using PyTorch's `@torch.jit.ignore` decorator.

If you continue to face challenges, sharing more specific details about the new module might help the community provide more targeted advice.

Good luck with your model export!

**Comment by djairjr** [comment]
Can I export just one category from the model, for example a person? Does this optimize tflite output?

**Reply by glenn-jocher** [reply]
@djairjr hello!

Thank you for your question. Yes, you can export a model to include only specific categories, such as "person," which can help optimize the TensorFlow Lite (TFLite) output by reducing the model size and potentially improving inference speed.

To achieve this, you can modify the model's class list to include only the desired categories before exporting. Here’s a step-by-step guide on how to do this:

1. **Filter Classes During Inference**: You can filter the classes during inference by specifying the class indices you want to keep. For example, if "person" is class 0, you can filter out other classes during inference.

2. **Custom Export Script**: Modify the export script to include only the desired classes. Here’s an example of how you can do this in Python:

    ```python
    from ultralytics import YOLO

    # Load the model
    model = YOLO("yolov8n.pt")  # or your custom model path

    # Define the classes to keep (e.g., 'person' class index is 0)
    keep_classes = [0]

    # Filter the model's class names
    model.names = {i: model.names[i] for i in keep_classes}

    # Export the model to TFLite format
    model.export(format="tflite")
    ```

3. **Post-Processing**: If you want to ensure that only the desired classes are detected during inference, you can add a post-processing step to filter out unwanted classes.

    ```python
    def filter_classes(results, keep_classes):
        filtered_results = []
        for result in results:
            if result['class'] in keep_classes:
                filtered_results.append(result)
        return filtered_results

    # Example usage during inference
    results = model("path/to/image.jpg")
    filtered_results = filter_classes(results, keep_classes=[0])
    ```

This approach ensures that your exported model is optimized for detecting only the specified categories, which can lead to better performance on edge devices.

For more detailed information on exporting models and the various formats supported, you can refer to our [Export Documentation](https://docs.ultralytics.com/integrations/).

If you encounter any issues or need further assistance, please provide a minimum reproducible example of your code, and ensure you are using the latest versions of `torch` and `ultralytics`. This will help us investigate and provide a more accurate solution.

Happy coding! 🚀

**Reply by djairjr** [reply]
I'm working on a project that detects if my dog ​​is peeing where he shouldn't and I trained a model in roboflow to detect pee stains. I would like to do something like the one proposed in this link: [https://y-t-g.github.io/tutorials/yolov8n-add-classes/](https://y-t-g.github.io/tutorials/yolov8n-add-classes/) that is, from Yolo's pre-trained models, I would like to select only the dog class . Then generate a final model that contains pee and dog. How do I do this? It's possible?

**Reply by glenn-jocher** [reply]
Hello @djairjr,

Thank you for reaching out! It's great to hear about your project. Yes, it is possible to create a model that detects both "dog" and "pee" by combining classes from a pre-trained YOLO model with your custom-trained model.

Here's a step-by-step guide to achieve this:

1. **Train Custom Model for Pee Detection**:
   - You’ve already trained a model to detect pee stains using Roboflow. Ensure this model is saved and ready for use.

2. **Filter Pre-trained Model for Dog Class**:
   - Load a pre-trained YOLO model and filter it to include only the "dog" class. Here’s how you can do it:

    ```python
    from ultralytics import YOLO

    # Load the pre-trained YOLO model
    model = YOLO("yolov8n.pt")

    # Filter to keep only the 'dog' class (assuming 'dog' class index is 16 in COCO dataset)
    keep_classes = [16]
    model.names = {i: model.names[i] for i in keep_classes}

    # Save the filtered model
    model.save("yolov8n_dog.pt")
    ```

3. **Combine Models**:
   - You can combine the detections from both models during inference. Here’s an example of how to do this:

    ```python
    from ultralytics import YOLO

    # Load the models
    dog_model = YOLO("yolov8n_dog.pt")
    pee_model = YOLO("path/to/pee_model.pt")

    # Run inference on an image
    image_path = "path/to/image.jpg"
    dog_results = dog_model(image_path)
    pee_results = pee_model(image_path)

    # Combine results
    combined_results = dog_results + pee_results

    # Visualize combined results
    combined_results.show()
    ```

4. **Export Combined Model**:
   - If you need to export the combined model to TFLite or another format, you can do so after combining the models. Here’s an example for TFLite:

    ```python
    combined_model = YOLO("path/to/combined_model.pt")
    combined_model.export(format="tflite")
    ```

This approach allows you to leverage the strengths of both models and create a final model that detects both "dog" and "pee" stains.

For more details on exporting models and the various formats supported, you can refer to our [Export Documentation](https://docs.ultralytics.com/integrations/).

If you encounter any issues or need further assistance, please provide a minimum reproducible example of your code, and ensure you are using the latest versions of `torch` and `ultralytics`. This will help us investigate and provide a more accurate solution.

Happy coding! 🚀

**Reply by djairjr** [reply]
I am always getting this error: super().__setattr__(name, value)
AttributeError: property 'names' of 'YOLO' object has no setter

On this line: model.names = {i: model.names[i] for i in keep_classes}

**Reply by glenn-jocher** [reply]
Hello @djairjr,

Thank you for reaching out! It looks like you're encountering an `AttributeError` when trying to set the `names` attribute of the `YOLO` object. This error occurs because the `names` attribute is read-only and does not have a setter method.

To achieve your goal of filtering the model to include only the "dog" class, you can use a different approach. Instead of trying to set the `names` attribute directly, you can filter the detections during inference. Here’s how you can do it:

1. **Filter Classes During Inference**:
   - You can filter the classes during inference by specifying the class indices you want to keep. For example, if "dog" is class 16 in the COCO dataset, you can filter out other classes during inference.

2. **Combine Models**:
   - Load your custom model for pee detection and the pre-trained model for dog detection. Combine their results during inference.

Here’s an example of how you can do this:

```python
from ultralytics import YOLO

# Load the models
dog_model = YOLO("yolov8n.pt")  # Pre-trained model
pee_model = YOLO("path/to/pee_model.pt")  # Custom model

# Define the classes to keep (e.g., 'dog' class index is 16)
keep_classes = [16]

def filter_classes(results, keep_classes):
    filtered_results = []
    for result in results:
        if result['class'] in keep_classes:
            filtered_results.append(result)
    return filtered_results

# Run inference on an image
image_path = "path/to/image.jpg"
dog_results = dog_model(image_path)
pee_results = pee_model(image_path)

# Filter dog results
filtered_dog_results = filter_classes(dog_results, keep_classes)

# Combine results
combined_results = filtered_dog_results + pee_results

# Visualize combined results
combined_results.show()
```

This approach ensures that your final model detects both "dog" and "pee" stains without modifying the `names` attribute directly.

If you encounter any further issues, please ensure you are using the latest versions of `torch` and `ultralytics`. If the problem persists, providing a minimum reproducible example of your code will help us investigate further. You can find more information on creating a minimum reproducible example [here](https://docs.ultralytics.com/help/minimum_reproducible_example).

Happy coding! 🚀

**Comment by djairjr** [comment]
163 / 5.000
Your code wasn't working.
It was giving error super().__setattr__(name, value)
AttributeError: can't set attribute 'names'

But this solution here works:
```
from ultralytics import YOLO

# Load the model
model = YOLO("yolov8n.pt")  # or your custom model path

# Define the class to keep (e.g., 'dog' class index is 16 in COCO dataset, we map it to 0)
keep_classes = {16: 0}

# Create a new dict for the names with only the class to keep
new_names = {0: model.names[16]}

# Adjust the class indices in the model
model.model.names = new_names
model.model.nc = len(new_names)
model.model.model[-1].nc = len(new_names)
model.model.model[-1].names = new_names

# Export the model to TFLite format
model.export(format="tflite")
```

**Reply by pderrenger** [reply]
@djairjr thank you for sharing your solution! It's great to see that you found a workaround for the issue. To ensure we address the root cause, could you please provide a minimum reproducible example of the code that originally triggered the `AttributeError: can't set attribute 'names'`? This will help us investigate the problem more effectively.

Additionally, please confirm that you are using the latest versions of `torch` and `ultralytics`. If not, upgrading these packages might resolve the issue:

```bash
pip install --upgrade torch ultralytics
```

Your provided solution for exporting the model to TFLite is clear and helpful. For others who might encounter similar issues, here's a concise version of your code:

```python
from ultralytics import YOLO

# Load the model
model = YOLO("yolov8n.pt")  # or your custom model path

# Define the class to keep (e.g., 'dog' class index is 16 in COCO dataset, we map it to 0)
keep_classes = {16: 0}

# Create a new dict for the names with only the class to keep
new_names = {0: model.names[16]}

# Adjust the class indices in the model
model.model.names = new_names
model.model.nc = len(new_names)
model.model.model[-1].nc = len(new_names)
model.model.model[-1].names = new_names

# Export the model to TFLite format
model.export(format="tflite")
```

For more detailed information on exporting models to various formats, you can refer to the [Ultralytics Export Documentation](https://docs.ultralytics.com/integrations/).

If you have any further questions or need additional assistance, feel free to ask. We're here to help! 😊

**Comment by djairjr** [comment]
No module named 'tensorflow_lite_support'

**Reply by pderrenger** [reply]
@djairjr hello! It looks like you're encountering an issue with the `tensorflow_lite_support` module. This typically occurs when the necessary TensorFlow Lite dependencies are not installed.

To resolve this, please ensure you have the required packages installed. You can do this by running:

```bash
pip install tensorflow tensorflow-lite-support
```

If the issue persists, please verify that you are using the latest versions of `torch` and `ultralytics` by running:

```bash
pip install --upgrade torch ultralytics
```

If you continue to experience problems, could you please provide a minimum reproducible example of your code? This will help us better understand the context and provide a more accurate solution. You can find guidance on creating a minimum reproducible example here: [Minimum Reproducible Example](https://docs.ultralytics.com/help/minimum_reproducible_example).

For more detailed information on exporting models to various formats, you can refer to our [Ultralytics Integrations](https://docs.ultralytics.com/integrations/) page.

Thank you for your patience, and we look forward to helping you resolve this issue! 😊

**Comment by djairjr** [comment]
I understand that the solution presented will work in the desktop environment, but how to implement a solution of this type in an edge environment, for example? The solutions I've been studying (Edge Impulse, Sense Ai) only allow me to input a model. Ideally, there would be a combine, or merge, method, so that it would result in a combined network.
Your solution will actually perform two detections, one for each object. The output you indicate: combined_results = filtered_dog_results + pee_results
it is a list and therefore, it does not have the show method pointed out.

There is a huge list of users proposing this implementation, or needing it. The tutorial I followed indicates a hack (patch) in the library that would enable this solution, but I was unable to get the resulting model to work.

I will try to manually train my model in roboflow, using both classes. For now, is the possible solution.

**Reply by pderrenger** [reply]
@djairjr hi there,

Thank you for your detailed question! Implementing a combined detection model for edge environments can indeed be challenging, especially when the platforms you're working with only allow a single model input.

To address your concerns:

1. **Combining Models**: Unfortunately, there isn't a built-in method to merge two separate YOLO models directly within the Ultralytics framework. However, you can train a single model to detect multiple classes (e.g., dogs and pee) by labeling your dataset accordingly and training the model on this combined dataset. This way, you only need to deploy one model that can detect both objects.

2. **Edge Deployment**: For deploying on edge devices, you can export your trained YOLO model to various formats that are optimized for edge environments. Here are some formats you might consider:
   - **TensorRT**: For NVIDIA GPUs, providing significant speedups.
   - **TFLite**: For TensorFlow Lite, suitable for mobile and embedded devices.
   - **ONNX**: For a versatile format that can be used with many different inference engines.
   - **CoreML**: For Apple devices.

   You can export your model using the following command:
   ```bash
   yolo export model=path/to/best.pt format=onnx  # or tflite, engine, coreml, etc.
   ```

3. **Handling Combined Results**: If you need to handle combined results from multiple detections, you can process the results in your application code. Here's a simple example in Python:
   ```python
   from ultralytics import YOLO

   # Load your model
   model = YOLO('path/to/combined_model.pt')

   # Run inference
   results = model('path/to/image.jpg')

   # Process results
   for result in results:
       print(result.boxes)  # Bounding boxes
       print(result.names)  # Class names
       result.show()  # Display results
   ```

4. **Manual Training**: Training your model with both classes in a tool like Roboflow is a good approach. Ensure your dataset is well-labeled and balanced to achieve the best performance.

For more detailed guidance on exporting models and integrating them into edge environments, please refer to our [Integrations Documentation](https://docs.ultralytics.com/integrations/).

I hope this helps! If you encounter any specific issues during the process, please provide a minimum reproducible example, and ensure you're using the latest versions of `torch` and `ultralytics`. This will help us investigate and provide a more accurate solution.

**Comment by Rad-Monk** [comment]
Hi, I wanted to create a model for web_ui element detection, how much data should i gather so that the model can perform atleast on 80% accuracy, I will be creating a dataset of my own since i wasn't able to find any reliable existing datasets.
I also created a web_ui_element classifier model from scratch, idk if it can help the detector model
Please let me know how i should go about it, I am new to this field, any advice would be useful.

**Reply by pderrenger** [reply]
@Rad-Monk hi there!

Creating a custom dataset for web UI element detection is a great approach, especially when existing datasets don't meet your needs. Here are some steps and tips to help you get started:

### Data Collection
1. **Diversity**: Ensure your dataset covers a wide range of web UI elements, including buttons, text fields, images, and other interactive components. This diversity will help your model generalize better.
2. **Quantity**: While there's no strict rule, starting with at least 500-1000 annotated images can provide a good baseline. More data generally leads to better performance, so aim to expand your dataset over time.
3. **Annotation**: Use tools like [LabelImg](https://github.com/tzutalin/labelImg) or [Roboflow](https://roboflow.com/) for annotating your images. Ensure that annotations are accurate and consistent.

### Model Training
1. **Pretrained Models**: Starting with a pretrained YOLO model (e.g., `yolov8n.pt`) can significantly speed up training and improve performance. Fine-tuning a pretrained model on your dataset is recommended.
2. **Training**: Here's a simple example to get you started with training:
    ```python
    from ultralytics import YOLO

    # Load a pretrained YOLO model
    model = YOLO("yolov8n.pt")

    # Train the model on your custom dataset
    results = model.train(data="path/to/your_dataset.yaml", epochs=50)
    ```

### Validation
1. **Validation Set**: Ensure you have a separate validation set to monitor the model's performance and avoid overfitting.
2. **Evaluation**: After training, evaluate your model to check its accuracy and make necessary adjustments:
    ```python
    results = model.val()
    ```

### Using Your Classifier
If you have a web UI element classifier, you can use it to pre-process images or as an additional feature in your detection pipeline. However, the primary focus should be on gathering a robust dataset and fine-tuning your detection model.

### Resources
For more detailed guidance, you can refer to the [Ultralytics Python Usage Documentation](https://docs.ultralytics.com/usage/python/).

Feel free to reach out if you have any more questions. Good luck with your project! 🚀

**Comment by rihablahouar** [comment]
Hello,

I am using a custom yolov8m.pt model for detection in a mobile application. When I run predictions using the YOLO CLI, the results are correct and accurate. However, after exporting the model to TFLite for deployment on the mobile app, the predictions are completely wrong on yolo CLI. What is the causes of this issues? 

model.export(format='saved_model', imgsz=640, optimize=True, keras=True)

I would like to know how to export the model to TFLite without losing accuracy and efficiency. Additionally, I would appreciate guidance on how to ensure the model predicts correctly with TensorFlowlite.

Thank you for your help.

**Reply by glenn-jocher** [reply]
It seems like the issue might be related to the export settings or model compatibility with TFLite. Ensure that the input size and quantization settings match your deployment requirements. You might also want to verify the model's accuracy post-export using a small test set. For more detailed guidance, please refer to the [Ultralytics export documentation](https://docs.ultralytics.com/modes/export/).

**Comment by avasanjesh** [comment]
Hi,

I have trained a detection yolov5 model. I export to onnx by command python export.py --weights best.pt --include onnx
when i run the inference with C++ it doesn't work. i tried to test it with python with command: python detect.py --weights best.onnx --source ..\data\images --conf-thres 0.25 --imgsz 640
but it gives me error which is :
detect: weights=['best.onnx'], source=..\data\images, data=data\coco128.yaml, imgsz=[640, 640], conf_thres=0.25, iou_thres=0.45, max_det=1000, device=, view_img=False, save_txt=False, save_format=0, save_csv=False, save_conf=Fal
se, save_crop=False, nosave=False, classes=None, agnostic_nms=False, augment=False, visualize=False, update=False, project=runs\detect, name=exp, exist_ok=False, line_thickness=3, hide_labels=False, hide_conf=False, half=False, dnn=False, vid_stride=1
YOLOv5  v7.0-420-g0c99ce80 Python-3.12.10 torch-2.7.0+cu126 CUDA:0 (NVIDIA GeForce RTX 3060 Laptop GPU, 6144MiB)

Loading best.onnx for ONNX Runtime inference...
Traceback (most recent call last):
  File "D:\Projects\Anomaly\YOLOv5Pooya\yolov5\detect.py", line 438, in <module>
    main(opt)
  File "D:\Projects\Anomaly\YOLOv5Pooya\yolov5\detect.py", line 433, in main
    run(**vars(opt))
  File "D:\Projects\.venv\Lib\site-packages\torch\utils\_contextlib.py", line 116, in decorate_context
    return func(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^
  File "D:\Projects\Anomaly\YOLOv5Pooya\yolov5\detect.py", line 252, in run
    s += f"{n} {names[int(c)]}{'s' * (n > 1)}, "  # add to string
                ~~~~~^^^^^^^^
KeyError: 1434


what is the problem? how to fix it? is the code for segmentation? mine is detection yolo5 model

**Reply by glenn-jocher** [reply]
@avasanjesh hello! The `KeyError: 1434` indicates that your exported ONNX model is predicting a class index (1434) that doesn't exist in your model's class names dictionary, which suggests an issue during the export process.

Here are a few steps to resolve this:

1. **Verify your original model works correctly:**
   ```bash
   python detect.py --weights best.pt --source ../data/images --conf-thres 0.25
   ```

2. **Check model class count and names before export:**
   ```python
   import torch
   model = torch.load('best.pt')
   print(f"Number of classes: {model['model'].nc}")
   print(f"Class names: {model['model'].names}")
   ```

3. **Re-export with explicit parameters:**
   ```bash
   python export.py --weights best.pt --include onnx --imgsz 640 --batch-size 1
   ```

4. **Validate the exported ONNX model:**
   ```bash
   python val.py --weights best.onnx --data your_dataset.yaml --imgsz 640
   ```

The issue likely stems from a mismatch between the model's expected class count and what's being predicted post-export. Make sure you're using the same dataset YAML file that was used during training, as this defines the class names and count.

If the problem persists, please ensure you're using compatible versions of PyTorch and ONNX, and consider re-training if the model file is corrupted.

**Reply by avasanjesh** [reply]
Thanks @glenn-jocher. I ran python detect.py --weights best.pt --source ../data/images --conf-thres 0.25  and i got error:
  File "D:\Projects\Anomaly\YOLOv5Pooya\yolov5\detect.py", line 252, in run
    s += f"{n} {names[int(c)]}{'s' * (n > 1)}, "  # add to string
                ~~~~~^^^^^^^^
KeyError: 234

but i infer correctly. the code is:

from ultralytics import YOLO
import os


# Load a model
model = YOLO("best.pt")  # pretrained YOLO11n model
h=0
output = 'InferResult29'
os.makedirs(output, exist_ok=True)
for i in os.listdir('2'):
    results = model(os.path.join('2',i))  # return a list of Results objects

    # Process results list
    for result in results:
        boxes = result.boxes  # Boxes object for bounding box outputs
        masks = result.masks  # Masks object for segmentation masks outputs
        keypoints = result.keypoints  # Keypoints object for pose outputs
        probs = result.probs  # Probs object for classification outputs
        obb = result.obb  # Oriented boxes object for OBB outputs
        # result.show()  # display to screen
        result.save(filename=f"{output}/{i.split('.')[0]}.jpg")
    h+=1

what is wrong? why detect.py doesnt work but my infer code works correctly?

**Reply by glenn-jocher** [reply]
@avasanjesh hello! The issue you're experiencing suggests a mismatch between the YOLOv5 repository's `detect.py` script and your model. Since your inference code using the Ultralytics YOLO class works correctly, but the YOLOv5 `detect.py` fails, this indicates you're mixing different frameworks.

The problem is that you're using the **YOLOv5 repository's detect.py** script with a model that was likely trained or saved using the **Ultralytics package** (which uses a different model format and class structure).

Here are two solutions:

**Option 1: Use Ultralytics CLI instead of YOLOv5 detect.py**
```bash
yolo predict model=best.pt source=../data/images conf=0.25 imgsz=640
```

**Option 2: Fix the class names issue in YOLOv5**
If you must use the YOLOv5 detect.py, ensure your model has the correct class names by checking:
```python
import torch
checkpoint = torch.load('best.pt', map_location='cpu')
print("Model keys:", checkpoint.keys())
if 'model' in checkpoint:
    print("Class names:", checkpoint['model'].names if hasattr(checkpoint['model'], 'names') else "No names found")
    print("Number of classes:", checkpoint['model'].nc if hasattr(checkpoint['model'], 'nc') else "No nc found")
```

**For ONNX export**, use the Ultralytics approach:
```bash
yolo export model=best.pt format=onnx imgsz=640
```

The KeyError suggests your model's class predictions don't match the expected class indices. Since your Ultralytics inference works, I recommend sticking with the Ultralytics framework for consistency.

**Reply by avasanjesh** [reply]
Thanks a lot. now onnx runs in python correctly with code:

from ultralytics import YOLO

onnx_model = YOLO("best.onnx")

# Run inference
results = onnx_model("../data/images/2_segment_28.png")

now i run the c++ code with this onnx model . i got the code from this page https://github.com/Hexmagic/ONNX-yolov5/blob/master/src/test.cpp I replace with my image and model but i get this error and i dont get any result image:


[ INFO:0@7.299] global onnx_importer.cpp:981 cv::dnn::dnn4_v20241223::ONNXImporter::handleNode DNN/ONNX: processing node with 5 inputs and 1 outputs: [Slice]:(onnx_node!/model.24/Slice_1) from domain='ai.onnx'
[ INFO:0@7.299] global onnx_importer.cpp:981 cv::dnn::dnn4_v20241223::ONNXImporter::handleNode DNN/ONNX: processing node with 2 inputs and 1 outputs: [Sub]:(onnx_node!/model.24/Sub) from domain='ai.onnx'
[ INFO:0@7.299] global onnx_importer.cpp:981 cv::dnn::dnn4_v20241223::ONNXImporter::handleNode DNN/ONNX: processing node with 2 inputs and 1 outputs: [Add]:(onnx_node!/model.24/Add_1) from domain='ai.onnx'
[ INFO:0@7.299] global onnx_importer.cpp:981 cv::dnn::dnn4_v20241223::ONNXImporter::handleNode DNN/ONNX: processing node with 2 inputs and 1 outputs: [Add]:(onnx_node!/model.24/Add_2) from domain='ai.onnx'
[ INFO:0@7.299] global onnx_importer.cpp:981 cv::dnn::dnn4_v20241223::ONNXImporter::handleNode DNN/ONNX: processing node with 2 inputs and 1 outputs: [Sub]:(onnx_node!/model.24/Sub_1) from domain='ai.onnx'
[ INFO:0@7.300] global onnx_importer.cpp:981 cv::dnn::dnn4_v20241223::ONNXImporter::handleNode DNN/ONNX: processing node with 2 inputs and 1 outputs: [Div]:(onnx_node!/model.24/Div_1) from domain='ai.onnx'
[ INFO:0@7.300] global onnx_importer.cpp:981 cv::dnn::dnn4_v20241223::ONNXImporter::handleNode DNN/ONNX: processing node with 2 inputs and 1 outputs: [Concat]:(onnx_node!/model.24/Concat_4) from domain='ai.onnx'
[ INFO:0@7.300] global onnx_importer.cpp:981 cv::dnn::dnn4_v20241223::ONNXImporter::handleNode DNN/ONNX: processing node with 2 inputs and 1 outputs: [Mul]:(onnx_node!/model.24/Mul_2) from domain='ai.onnx'
[ INFO:0@7.300] global onnx_importer.cpp:981 cv::dnn::dnn4_v20241223::ONNXImporter::handleNode DNN/ONNX: processing node with 2 inputs and 1 outputs: [Concat]:(onnx_node!/model.24/Concat_5) from domain='ai.onnx'
[ INFO:0@7.312] global registry_parallel.impl.hpp:96 cv::parallel::ParallelBackendRegistry::ParallelBackendRegistry core(parallel): Enabled backends(3, sorted by priority): ONETBB(1000); TBB(990); OPENMP(980)
[ INFO:0@7.313] global plugin_loader.impl.hpp:67 cv::plugin::impl::DynamicLib::libraryLoad load C:\Users\User\Downloads\Programs\opencv\build\x64\vc16\bin\opencv_core_parallel_onetbb4110_64d.dll => FAILED
[ INFO:0@7.314] global plugin_loader.impl.hpp:67 cv::plugin::impl::DynamicLib::libraryLoad load opencv_core_parallel_onetbb4110_64d.dll => FAILED
[ INFO:0@7.314] global plugin_loader.impl.hpp:67 cv::plugin::impl::DynamicLib::libraryLoad load C:\Users\User\Downloads\Programs\opencv\build\x64\vc16\bin\opencv_core_parallel_tbb4110_64d.dll => FAILED
[ INFO:0@7.314] global plugin_loader.impl.hpp:67 cv::plugin::impl::DynamicLib::libraryLoad load opencv_core_parallel_tbb4110_64d.dll => FAILED
[ INFO:0@7.315] global plugin_loader.impl.hpp:67 cv::plugin::impl::DynamicLib::libraryLoad load C:\Users\User\Downloads\Programs\opencv\build\x64\vc16\bin\opencv_core_parallel_openmp4110_64d.dll => FAILED
[ INFO:0@7.315] global plugin_loader.impl.hpp:67 cv::plugin::impl::DynamicLib::libraryLoad load opencv_core_parallel_openmp4110_64d.dll => FAILED
OpenCV(4.11.0) Error: Assertion failed (total(srcShape, srcRange.start, srcRange.end) == maskTotal) in cv::dnn::computeShapeByReshapeMask, file C:\GHA-OCV-1\_work\ci-gha-workflow\ci-gha-workflow\opencv\modules\dnn\src\layers\reshape_layer.cpp, line 109
[ERROR:0@7.328] global net_impl.cpp:1165 cv::dnn::dnn4_v20241223::Net::Impl::getLayerShapesRecursively OPENCV/DNN: [Reshape]:(onnx_node!/model.24/dfl/Reshape): getMemoryShapes() throws exception. inputs=1 outputs=1/1 blobs=0
[ERROR:0@7.329] global net_impl.cpp:1168 cv::dnn::dnn4_v20241223::Net::Impl::getLayerShapesRecursively     input[0] = [ 1 64 4200 ]
[ERROR:0@7.329] global net_impl.cpp:1172 cv::dnn::dnn4_v20241223::Net::Impl::getLayerShapesRecursively     output[0] = [ ]
[ERROR:0@7.329] global net_impl.cpp:1178 cv::dnn::dnn4_v20241223::Net::Impl::getLayerShapesRecursively Exception message: OpenCV(4.11.0) C:\GHA-OCV-1\_work\ci-gha-workflow\ci-gha-workflow\opencv\modules\dnn\src\layers\reshape_layer.cpp:109: error: (-215:Assertion failed) total(srcShape, srcRange.start, srcRange.end) == maskTotal in function 'cv::dnn::computeShapeByReshapeMask'

OpenCV: terminate handler is called! The last OpenCV error is:
OpenCV(4.11.0) Error: Assertion failed (total(srcShape, srcRange.start, srcRange.end) == maskTotal) in cv::dnn::computeShapeByReshapeMask, file C:\GHA-OCV-1\_work\ci-gha-workflow\ci-gha-workflow\opencv\modules\dnn\src\layers\reshape_layer.cpp, line 109

what is the problem? thanks in advance

**Reply by glenn-jocher** [reply]
@avasanjesh hello! The error you're encountering is due to OpenCV DNN not fully supporting the DFL (Distribution Focal Loss) operations in newer YOLO models. The ONNX model exported from Ultralytics YOLO (YOLOv8/YOLO11) contains operations that OpenCV DNN cannot handle properly.

Here are your options to resolve this:

**Option 1: Use ONNX Runtime instead of OpenCV DNN**
Replace OpenCV DNN with ONNX Runtime in your C++ code, which has better support for modern YOLO architectures.

**Option 2: Export without DFL (if supported)**
Try exporting with simplified operations:
```bash
yolo export model=best.pt format=onnx simplify=True
```

**Option 3: Use YOLOv5 for OpenCV DNN compatibility**
Since you're using C++ code designed for YOLOv5, consider training your model with the YOLOv5 repository directly, as YOLOv5 models are more compatible with OpenCV DNN.

**Option 4: Use TensorRT or other inference engines**
For better performance and compatibility, consider exporting to TensorRT:
```bash
yolo export model=best.pt format=engine
```

The OpenCV DNN + YOLO combination works best with YOLOv5 models. For YOLOv8/YOLO11 models, ONNX Runtime or TensorRT are recommended inference engines.

**Reply by avasanjesh** [reply]
Thanks. I tried options 2 to 4 but they didnt work. can you guide me how i can install onnx on microsoft visual studio? how does the c++ code changes to use this onnx instead of opencv? i use the code written on your page. https://github.com/Hexmagic/ONNX-yolov5/blob/master/src/test.cpp

**Reply by glenn-jocher** [reply]
@avasanjesh hello! Here's how to set up ONNX Runtime with Visual Studio for your C++ project:

**Installing ONNX Runtime:**

1. **Download ONNX Runtime C++ package:**
   - Go to [ONNX Runtime releases](https://github.com/microsoft/onnxruntime/releases)
   - Download `Microsoft.ML.OnnxRuntime.{version}.zip` for Windows

2. **Visual Studio Setup:**
   - Extract the downloaded package
   - In your VS project properties:
     - Add `include` folder to Additional Include Directories
     - Add `lib` folder to Additional Library Directories  
     - Add `onnxruntime.lib` to Additional Dependencies
     - Copy `onnxruntime.dll` to your output directory

**Code Changes:**
The current OpenCV DNN code won't work directly. You'll need to replace the OpenCV DNN inference with ONNX Runtime. Here's a basic structure:

```cpp
#include <onnxruntime_cxx_api.h>

// Replace cv::dnn::Net with Ort::Session
Ort::Session session{env, model_path, session_options};

// Replace net.setInput() and net.forward() with:
// - Create input tensor
// - Run inference: session.Run()
// - Process output tensors
```

**Alternative Solution:**
Since modifying the entire inference code is complex, I recommend using existing ONNX Runtime + YOLO C++ examples specifically designed for YOLOv8/YOLO11:

- [YOLOv8 ONNX Runtime C++](https://github.com/UNeedCryDear/yolov8-opencv-onnxruntime-cpp)
- [YOLO ONNX Runtime Examples](https://github.com/DefTruth/lite.ai.toolkit)

These repositories provide complete working examples with ONNX Runtime integration that should work with your exported model.
